rpc: add tcp_rtt and tcp_rtt_var metrics for gRPC

xuchef opened the pull request:
### `util/sysutil: add functions for getting TCP round-trip time and variance`
Previously, we relied on timing heartbeats in the RPC layer to estimate network
latency (rpc.connection.avg_round_trip_latency). However, because these
measurements are computed within cockroach, they can be confounded by CPU-heavy
workloads.

Luckily, Linux maintains the smoothed round-trip time (SRTT) and RTT variance
for each TCP socket it opens. As kernel-computed metrics, these are less
sensitive to CPU overload.

This patch adds a sysutil function for obtaining SRTT and RTT variance given a
a net.TCPConn pointer.

Part of: https://github.com/cockroachdb/cockroach/issues/149959
Release note: None

### `rpc: add tcp_rtt and tcp_rtt_var metrics for gRPC`

Previously, our only metric for gauging network latency was
rpc.connection.avg_round_trip_latency. This metric was calculated by timing
heartbeats in the RPC layer. However, because these measurements are computed
within cockroach, they can be confounded by CPU-heavy workloads.

Through escalations, we've found that elevated network latencies (outside of
CRDB's control) can severely degrade cluster performance. So, being able to
directly and accurately identify these cases would be helpful.

To address this, this patch introduces two new metrics whose values are computed
by Linux. As kernel-computed metrics, these are less sensitive to CPU overload:
1. rpc.connection.tcp_rtt: TCP smoothed round-trip time
2. rpc.connection.tcp_rtt_var: TCP round-trip time variance

Since these metrics are internally aggregated by Linux, we only need to sample
them periodically. We update them in the heartbeat loop, at the same cadence as
the original avg_round_trip_latency.

To obtain these metrics, we need access to the underlying *net.TCPConn of our
gRPC peer connection. So, the dial function we pass to gRPC has been modified to
update the tcpConn field of the peer struct on each network dial.

Part of: https://github.com/cockroachdb/cockroach/issues/149959
Release note: None

xuchef committed:
util/sysutil: add functions for getting TCP round-trip time and variance

Previously, we relied on timing heartbeats in the RPC layer to estimate network
latency (rpc.connection.avg_round_trip_latency). However, because these
measurements are computed within cockroach, they can be confounded by CPU-heavy
workloads.

Luckily, Linux maintains the smoothed round-trip time (SRTT) and RTT variance
for each TCP socket it opens. As kernel-computed metrics, these are less
sensitive to CPU overload.

This patch adds a sysutil function for obtaining SRTT and RTT variance given a
a net.TCPConn pointer.

Part of: #149959
Release note: None

xuchef commented on pkg/rpc/grpc.go:
```diff
@@ -63,7 +64,24 @@ func newGRPCPeerOptions(
+					defer p.mu.Unlock()
+					p.mu.tcpConn = tcpConn
+				}
+				return rpcCtx.grpcDialRaw(ctx, target, class, onNetworkDial, additionalDialOpts...)
```
I don't think this solution is particularly clean, so I'd be happy to hear if you have any suggestions on how this could be made better.

I think this is mainly complicated by how we've got two code paths that need these gRPC options.

<img width="1412" height="604" alt="Image" src="https://github.com/user-attachments/assets/44c8fabc-df6a-4b09-bf7e-de5b73b04589" />

We explored a couple ideas for how to go about this:

1. We considered storing and updating the metrics on the connection level. See [this branch](https://github.com/cockroachdb/cockroach/compare/master...xuchef:cockroach:tcp-rtt-wrapped-net-conn#diff-335c971a96e15ab11ca9738f2f81feabc5b5209b87209024b1015b95e194edac) for a quick prototype of what this would look like. Architecturally this makes a lot of sense as these metrics are after all on a connection level. But practically, I don't think it's what we want because we'd then have to either update the metrics on write (introducing latency for not much value) or delegate that back to the heartbeat loop (would be simpler if we just stored the conn on the peer). So, as you mentioned in your [comment about connections being fungible](https://github.com/cockroachdb/cockroach/issues/149959#issuecomment-3088753045), it'd be much simpler if we did it the way it is here, where we [store a fungible, representative tcpConn](https://github.com/cockroachdb/cockroach/pull/150869/files#diff-fc54146a422d3d59215aae7d10ecb55771db57f0f49b0b0fa34cd64358979ca0R205-R210).
2. Rather than passing this onNetworkDial callback, set the dialer func here directly (i.e. move [this code](https://github.com/xuchef/cockroach/blob/b3659b3b1df63b7ab6ff2699541bd0a8519f66c3/pkg/rpc/context.go#L1628-L1654) up here). I chose the callback over this approach because I think it makes sense for the dialer to be set in `dialOptsNetwork`.

xuchef reviewed (commented):
Hey @tbg I realize that the numbers from this morning were way higher than expected because the graph was including the aggregated per-node metrics (now filtered out with `rpc_connection_tcp_rtt{class!=""}`). This looks better:

<img width="2096" height="1806" alt="Image" src="https://github.com/user-attachments/assets/01e7c542-a725-4215-b049-f03120b18d0b" />

mgartner commented:
Should this close #149959, or should that issue only close once we collect the same metrics with DRPC too?

xuchef commented:
# Testing

## Setup

First, I applied the following patch to log when we're dialing a TCP connection. Not committed for now cause it might be noisy, but happy to add it if it'll be useful in the future.
```diff
diff --git a/pkg/rpc/grpc.go b/pkg/rpc/grpc.go
index d5c712f52a6..cfe86d71afd 100644
--- a/pkg/rpc/grpc.go
+++ b/pkg/rpc/grpc.go
@@ -12,6 +12,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/rpc/rpcbase"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/connectivity"
@@ -80,6 +81,8 @@ func newGRPCPeerOptions(
 					p.mu.Lock()
 					defer p.mu.Unlock()
 					p.mu.tcpConn = tcpConn
+
+					log.Infof(ctx, "gRPC network dial: laddr=%v", tcpConn.LocalAddr())
 				}
 				return rpcCtx.grpcDialRaw(ctx, target, class, onNetworkDial, additionalDialOpts...)
 			},

```

Using this patch, I built the binary.
```bash
./dev build --cross cockroach
```

And then started a roachprod cluster with it.
```bash
c=michaelxu-tcp-rtt-metrics

roachprod create $c -n 6
roachprod put $c ./artifacts/cockroach
roachprod start $c
```

I turned on per-connection class metrics.
```bash
roachprod sql $c:6 -- -e 'SET CLUSTER SETTING server.child_metrics.enabled = true;'
```

After letting the cluster warm up, I ran a kv workload. See [here](https://github.com/cockroachdb/cockroach/pull/149928#issuecomment-3075394996) for more details.
```bash
roachprod ssh $c:6 -- "
./cockroach workload init kv \
	--drop \
	--scatter \
	--splits 500 \
	--key-size 1024 \
	{pgurl:1-5}
"

roachprod ssh $c:6 -- "
./cockroach workload run kv \
	--key-size 1024 \
	--read-percent 95  \
	--max-rate 250 \
	--batch 100 \
	{pgurl:1-5}
"
```

## Initial Results
For simplicity, we'll focus on connections from `node 1` -> `node 4`.

Full results are persisted on the [testeng grafana instance](https://grafana.testeng.crdb.io/d/aetery46nxji8e/tcp-rtt?orgId=1&from=1754951291429&to=1754952190574&var-DS_PROMETHEUS=v9Zz2K6nz&var-node=1&var-rate_interval=30s&var-from=1&var-to=2&var-class=system&var-cluster=michaelxu-tcp-rtt-metrics&var-filters=node_id%7C%3D%7C1&var-filters=remote_node_id%7C%3D%7C4).

### Network Dials
As expected, we have three TCP network dials from `n1` -> `n4`, one for each [connection class](https://github.com/cockroachdb/cockroach/blob/fb15a49058e015e1fc104e8884292c734975ccd6/pkg/rpc/rpcbase/connection_class.go#L37-L49). Note that we don't have the rangefeed connection class because we don't have rangefeeds enabled.
```bash
$ roachprod ssh $c:1 -- "cat logs/cockroach.log | grep 'n1,rnode=4.*gRPC network dial'"
I250811 22:26:52.103637 4305 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=4,raddr=‹10.142.0.122:26257›,class=system,rpc] 175  gRPC network dial: laddr=‹10.142.0.35:44592›
I250811 22:26:52.134404 4713 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=4,raddr=‹10.142.0.122:26257›,class=default,rpc] 177  gRPC network dial: laddr=‹10.142.0.35:44594›
I250811 22:26:52.172847 4869 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=4,raddr=‹10.142.0.122:26257›,class=raft,rpc] 179  gRPC network dial: laddr=‹10.142.0.35:44602›
```

### Metric Graphs

At the top of our grafana dashboard, we can add a filter for `n1` -> `n4`.
<img width="2362" height="120" alt="Screenshot 2025-08-11 at 2 28 34 PM" src="https://github.com/user-attachments/assets/2ddf6ca5-4045-4b6c-b2b7-c41470e5ef06" />

We have the following graphs:

`rpc_connection_avg_round_trip_latency`
- existing RPC RTT metric
- a timer in the cockroach process times how long hearbeat RPCs take
	- updated on each heartbeat => [once per second](https://github.com/cockroachdb/cockroach/blob/bcc1685671914fd8200b32c753511aa233283727/pkg/base/config.go#L165)
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/a3bbe2a7-5f23-43b1-827f-7e41c85591a4" />

---

`rpc_connection_tcp_rtt`
- *newly added metric*
- kernel-computed TCP RTT metric, smoothed with an EWMA (see [RFC 6298](https://datatracker.ietf.org/doc/html/rfc6298#section-2) for more details)
	- updated using TCP timestamps (`TSval`, `TSecr`) on each packet ACK
- *sampled by us* on each heartbeat => once per second
	- so this is missing resolution
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/671a9291-4dc0-4577-a249-9a74a84bfd36" />

---

`rpc_connection_tcp_rtt_var`
- *newly added metric*
- kernel-computed TCP RTT Variance metric, not technically variance, but an estimate of it (see [RFC 6298](https://datatracker.ietf.org/doc/html/rfc6298#section-2) for more details)
	- updated using TCP timestamps (`TSval`, `TSecr`) on each packet ACK
- *sampled by us* on each heartbeat => once per second
	- so this is missing resolution
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/36bf5c3f-0906-4318-8b84-441e06b7b928" />

### Analysis
In the steady state, we're seeing that TCP RTT is consistently greater than the heartbeat RTT. 

At first glance, I thought something was wrong: as the kernel metrics are being recorded from a lower level, I initially expected them to be strictly less than the heartbeat latencies. But! the TCP connection handles wayyy more RPCs than just the heartbeart (gRPC uses HTTP2, so multiplexes multiple gRPC streams over an individual TCP connection). The TCP RTT is much higher because it accounts for *all* traffic through the TCP connection—the heartbeat RPC is a subset of this.

I'm not quite sure why the TCP RTT is so spiky. But my best guess is that we have lots of things that are periodic and that's causing the spikiness pattern that we're seeing: when there's traffic, it jumps up; when there isn't it calms back down.

To explore this further, I turned to https://github.com/iovisor/bcc/blob/master/tools/tcprtt.py. The `tcprtt` tool from the BCC toolkit attaches a kprobe to the `tcp_rcv_established` event, and on each event, samples the smoothed TCP RTT. That is, it does the same thing our new metrics do, except at a much higher frequency, on every TCP segment received.

I installed the BCC toolkit from [source](https://github.com/iovisor/bcc/blob/master/INSTALL.md#ubuntu---source) because apparently the Ubuntu [binaries are outdated](https://github.com/iovisor/bcc/blob/master/INSTALL.md#ubuntu---binary).
```bash
# https://github.com/iovisor/bcc/blob/master/INSTALL.md#install-build-dependencies-1
roachprod ssh $c -- "
  sudo apt install -y zip bison build-essential cmake flex git libedit-dev \
    libllvm14 llvm-14-dev libclang-14-dev python3 zlib1g-dev libelf-dev libfl-dev python3-setuptools \
    liblzma-dev libdebuginfod-dev arping netperf iperf \
    python-is-python3
"

# https://github.com/iovisor/bcc/blob/master/INSTALL.md#install-bcc-from-source
roachprod ssh $c -- "
  git clone https://github.com/iovisor/bcc.git
  mkdir bcc/build; cd bcc/build
  cmake ..
  make
  sudo make install
  cmake -DPYTHON_CMD=python3 ..
  pushd src/python/
  make
  sudo make install
  popd
"
```
Now, `tcprtt` (along with all the other BCC tools) are installed in `/usr/share/bcc/tools`.

From above, we saw that the local address for the raft client connection for `n1`->`n4` is `10.142.0.35:44602`.
> I250811 22:26:52.172847 4869 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=4,raddr=‹10.142.0.122:26257›,class=raft,rpc] 179  gRPC network dial: laddr=‹10.142.0.35:44602›

Running `tcprtt` for 60 seconds in 15 second intervals yields on this socket:
```bash
$ roachprod ssh $c:1 -- sudo /usr/share/bcc/tools/tcprtt -b -T -i 15 -d 60 -a 10.142.0.35 -p 44602
Tracing TCP RTT... Hit Ctrl-C to end.

22:50:41

Local Address = 10.142.0.35
     usecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 189      |**********                              |
       128 -> 255        : 690      |****************************************|
       256 -> 511        : 392      |**********************                  |
       512 -> 1023       : 432      |*************************               |
      1024 -> 2047       : 411      |***********************                 |
      2048 -> 4095       : 232      |*************                           |
      4096 -> 8191       : 127      |*******                                 |
      8192 -> 16383      : 4        |                                        |

22:50:56

Local Address = 10.142.0.35
     usecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 150      |******                                  |
       128 -> 255        : 919      |****************************************|
       256 -> 511        : 436      |******************                      |
       512 -> 1023       : 364      |***************                         |
      1024 -> 2047       : 344      |**************                          |
      2048 -> 4095       : 223      |*********                               |
      4096 -> 8191       : 104      |****                                    |
      8192 -> 16383      : 2        |                                        |

22:51:11

Local Address = 10.142.0.35
     usecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 234      |***********                             |
       128 -> 255        : 849      |****************************************|
       256 -> 511        : 407      |*******************                     |
       512 -> 1023       : 358      |****************                        |
      1024 -> 2047       : 402      |******************                      |
      2048 -> 4095       : 253      |***********                             |
      4096 -> 8191       : 107      |*****                                   |

22:51:26

Local Address = 10.142.0.35
     usecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 165      |*******                                 |
       128 -> 255        : 827      |****************************************|
       256 -> 511        : 400      |*******************                     |
       512 -> 1023       : 298      |**************                          |
      1024 -> 2047       : 275      |*************                           |
      2048 -> 4095       : 196      |*********                               |
      4096 -> 8191       : 122      |*****                                   |
      8192 -> 16383      : 2        |                                        |
```

Matching this up with the [our graphs](https://grafana.testeng.crdb.io/d/aetery46nxji8e/tcp-rtt?orgId=1&from=1754952641000&to=1754952705000&var-DS_PROMETHEUS=v9Zz2K6nz&var-node=1&var-rate_interval=30s&var-from=1&var-to=2&var-class=system&var-cluster=michaelxu-tcp-rtt-metrics&var-filters=node_id%7C%3D%7C1&var-filters=remote_node_id%7C%3D%7C4).
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/6614996d-03fd-41be-b427-d015c4301259" />
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/121765e2-3f86-44b4-ba44-be9a7d3fe581" />
<img width="3014" height="578" alt="image" src="https://github.com/user-attachments/assets/7dd67add-23a3-49fa-aabd-d47805fed039" />

I seems that our metrics check out: they are displaying the right thing, just at a much lower resolution. This shows that the underlying TCP RTT is actually highly variable within a 1s window. And by nature of our 1s sampling rate, we see these low-resolution steps in the graph.

As to why there is so much variability, I'm not quite sure.

## Subsequent Experiments

### Adding 50ms Packet Delay

We can artificially inject 50ms delay to all egress traffic from `n4`. We'd expect both of our RTT metrics to shoot up to 50ms.
```bash
roachprod ssh $c:4 -- sudo tc qdisc add dev ens5 root netem delay 50ms
```

As expected, this increases both RTT metrics to the ~50ms range ([graph](https://grafana.testeng.crdb.io/d/aetery46nxji8e/tcp-rtt?orgId=1&from=1754953306763&to=1754953530065&var-DS_PROMETHEUS=v9Zz2K6nz&var-node=1&var-rate_interval=30s&var-from=1&var-to=2&var-class=system&var-cluster=michaelxu-tcp-rtt-metrics&var-filters=node_id%7C%3D%7C1&var-filters=remote_node_id%7C%3D%7C4)).
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/e439ea7e-9988-407b-95f1-3dfd6d55c8de" />
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/bcf19dd8-1ee2-4f6c-a227-a5f6e844f3e2" />

This provides evidence that this new metric is measuring the right thing (and in the right units!).

After letting it run for a bit, we can remove this rule.
```bash
roachprod ssh $c:4 -- sudo tc qdisc del dev ens5 root
```

### Adding 10% Packet Loss
Repeating the experiment from [Tobi's TCP retransmission metrics PR](https://github.com/cockroachdb/cockroach/pull/149928#issuecomment-3075394996), we'll add 10% packet loss between all nodes in the cluster.
```bash
roachprod ssh $c -- sudo sysctl -w net.ipv4.tcp_early_retrans=0
roachprod ssh $c:1-5 -- sudo iptables -A OUTPUT -p tcp -m tcp --dport 26257 -m statistic --mode random --probability 0.1 -j DROP
```

We'd expect this to have an impact on the pre-existing `rpc_connection_tcp_rtt` metric because packet loss will increase the number of TCP retransmissions, thereby elevating gRPC latency. However, we would *not* expect to see an impact on our newly added metrics because they are only updated upon receiving an ACK—and we won't receive ACKs for dropped packets. See Experiment 1 in [this paper](https://queue.acm.org/detail.cfm?id=2539132) for more details.

As expected, we see elevated heartbeat RPC latencies.
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/c0579e79-4fc9-4147-b96e-6d24ffd52ce0" />

As expected, we see relatively no change to TCP RTT.
<img width="3014" height="600" alt="image" src="https://github.com/user-attachments/assets/89b9ddaf-6f15-45db-a50b-062cdb9f777a" />

After letting it run for a bit, we can restore the iptables.
```bash
roachprod ssh $c:1-5 -- "sudo iptables-save  | grep -v 26257 | sudo iptables-restore"
```

xuchef committed:
rpc: add tcp_rtt and tcp_rtt_var metrics for gRPC

Previously, our only metric for gauging network latency was
rpc.connection.avg_round_trip_latency. This metric was calculated by timing
heartbeats in the RPC layer. However, because these measurements are computed
within cockroach, they can be confounded by CPU-heavy workloads.

Through escalations, we've found that elevated network latencies (outside of
CRDB's control) can severely degrade cluster performance. So, being able to
directly and accurately identify these cases would be helpful.

To address this, this patch introduces two new metrics whose values are computed
by Linux. As kernel-computed metrics, these are less sensitive to CPU overload:
1. rpc.connection.tcp_rtt: TCP smoothed round-trip time
2. rpc.connection.tcp_rtt_var: TCP round-trip time variance

Since these metrics are internally aggregated by Linux, we only need to sample
them periodically. We update them in the heartbeat loop, at the same cadence as
the original avg_round_trip_latency.

To obtain these metrics, we need access to the underlying *net.TCPConn of our
gRPC peer connection. So, the dial function we pass to gRPC has been modified to
update the tcpConn field of the peer struct on each network dial.

Part of: cockroachdb#149959
Release note: None

cthumuluru-crdb commented on pkg/util/sysutil/tcpinfo_stub.go:
```diff
@@ -0,0 +1,15 @@
+import "syscall"
+
+// GetRTTInfo is a stub implementation for non-Linux platforms returning (nil, false).
+func GetRTTInfo(conn syscall.RawConn) (value *RTTInfo, ok bool) {
```
Shouldn't this be `func GetRTTInfo(conn *net.TCPConn) (value *RTTInfo, ok bool)` ?

cthumuluru-crdb reviewed (commented):
I left a couple of minor comments. Otherwise the changes look good.

cthumuluru-crdb reviewed (approved):
lgtm!

xuchef commented on pkg/rpc/grpc.go:
```diff
@@ -63,7 +64,24 @@ func newGRPCPeerOptions(
+
+					p.mu.Lock()
+					defer p.mu.Unlock()
+					p.mu.tcpConn = tcpConn
```
Sounds good. I was working on the DRPC stuff and opened this PR for it https://github.com/cockroachdb/cockroach/pull/151743

xuchef commented:
TFTR!

I added a verbose log line 
https://github.com/cockroachdb/cockroach/blob/3d8c5ca253a4539145290b61e1e93539c9511342/pkg/rpc/grpc.go#L85

Turning on verbose logs for `grpc.go`
```bash
roachprod start $c --args="--vmodule=grpc=2"
```

I get
```bash
$ rg 'gRPC network dial' ~/local/1
/home/michael_xu_cockroachlabs_com/local/1/logs/cockroach.gceworker-mxu.michaelxucockroachlabscom.2025-08-12T18_08_19Z.1059307.log
81:I250812 18:08:20.344908 775 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=?,raddr=‹127.0.0.1:29000›,class=system,rpc] 42  gRPC network dial: laddr=‹127.0.0.1:35050›
215:I250812 18:08:21.252894 2804 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=2,raddr=‹gceworker-mxu:29002›,class=system,rpc] 161  gRPC network dial: laddr=‹10.142.0.77:48346›
217:I250812 18:08:21.296208 2726 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=2,raddr=‹gceworker-mxu:29002›,class=default,rpc] 163  gRPC network dial: laddr=‹10.142.0.77:48366›
219:I250812 18:08:21.347132 2936 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=?,raddr=‹gceworker-mxu:29002›,class=system,rpc] 165  gRPC network dial: laddr=‹10.142.0.77:48376›
223:I250812 18:08:22.075421 3044 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=3,raddr=‹gceworker-mxu:29004›,class=system,rpc] 169  gRPC network dial: laddr=‹10.142.0.77:54964›
226:I250812 18:08:22.097419 3098 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=3,raddr=‹gceworker-mxu:29004›,class=default,rpc] 172  gRPC network dial: laddr=‹10.142.0.77:55000›
229:I250812 18:08:22.143697 2911 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=2,raddr=‹gceworker-mxu:29002›,class=raft,rpc] 175  gRPC network dial: laddr=‹10.142.0.77:48392›
231:I250812 18:08:22.242337 3382 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=3,raddr=‹gceworker-mxu:29004›,class=raft,rpc] 177  gRPC network dial: laddr=‹10.142.0.77:55008›
234:I250812 18:08:22.963146 3729 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=4,raddr=‹gceworker-mxu:29006›,class=system,rpc] 180  gRPC network dial: laddr=‹10.142.0.77:45672›
237:I250812 18:08:23.000247 3887 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=4,raddr=‹gceworker-mxu:29006›,class=raft,rpc] 183  gRPC network dial: laddr=‹10.142.0.77:45708›
238:I250812 18:08:23.000348 3816 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=4,raddr=‹gceworker-mxu:29006›,class=default,rpc] 184  gRPC network dial: laddr=‹10.142.0.77:45692›
242:I250812 18:08:23.860172 4623 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=5,raddr=‹gceworker-mxu:29008›,class=system,rpc] 188  gRPC network dial: laddr=‹10.142.0.77:43770›
245:I250812 18:08:23.910059 4640 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=5,raddr=‹gceworker-mxu:29008›,class=raft,rpc] 191  gRPC network dial: laddr=‹10.142.0.77:43786›
246:I250812 18:08:23.910194 3928 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=5,raddr=‹gceworker-mxu:29008›,class=default,rpc] 192  gRPC network dial: laddr=‹10.142.0.77:43778›
251:I250812 18:08:24.743939 5830 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=6,raddr=‹gceworker-mxu:29010›,class=system,rpc] 197  gRPC network dial: laddr=‹10.142.0.77:60324›
254:I250812 18:08:24.797485 6044 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=6,raddr=‹gceworker-mxu:29010›,class=default,rpc] 200  gRPC network dial: laddr=‹10.142.0.77:60382›
256:I250812 18:08:24.826672 6210 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=6,raddr=‹gceworker-mxu:29010›,class=raft,rpc] 202  gRPC network dial: laddr=‹10.142.0.77:60392›
1165:I250812 18:08:36.459979 28451 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=3,raddr=‹gceworker-mxu:29004›,class=rangefeed,rpc] 1111  gRPC network dial: laddr=‹10.142.0.77:37620›
1166:I250812 18:08:36.460212 28407 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=4,raddr=‹gceworker-mxu:29006›,class=rangefeed,rpc] 1112  gRPC network dial: laddr=‹10.142.0.77:59894›
1169:I250812 18:08:36.477566 28535 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=2,raddr=‹gceworker-mxu:29002›,class=rangefeed,rpc] 1115  gRPC network dial: laddr=‹10.142.0.77:45804›
1171:I250812 18:08:36.506123 28507 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=6,raddr=‹gceworker-mxu:29010›,class=rangefeed,rpc] 1117  gRPC network dial: laddr=‹10.142.0.77:44574›
1172:I250812 18:08:36.542241 28427 rpc/grpc.go:85 ⋮ [T1,Vsystem,n1,rnode=5,raddr=‹gceworker-mxu:29008›,class=rangefeed,rpc] 1118  gRPC network dial: laddr=‹10.142.0.77:50148›
```
